gh-91081: Add note on WeakKeyDictionary behavior when deleting a replaced entry (#91499)
python/cpython: Doc/library/weakref.rst

   application without adding attributes to those objects.  This can be especially
   useful with objects that override attribute accesses.

   Note that when a key with equal value to an existing key (but not equal identity)
   is inserted into the dictionary, it replaces the value but does not replace the
   existing key. Due to this, when the reference to the original key is deleted, it
   also deletes the entry in the dictionary::

>       d = WeakKeyDictionary()

Reviewer: Can you add a definition of `k1` and `k2`  here? Without it, this example is hard to understand.
```suggestion
      class T(str):
          pass
      k1 = T('hello')
      k2 = T('hello')
      d = WeakKeyDictionary()
```

   application without adding attributes to those objects.  This can be especially
   useful with objects that override attribute accesses.

   Note that when a key with equal value to an existing key (but not equal identity)
   is inserted into the dictionary, it replaces the value but does not replace the
   existing key. Due to this, when the reference to the original key is deleted, it
   also deletes the entry in the dictionary::

      >>> class T(str):
      ...     pass
      ...
      >>> k1, k2 = T(), T()
      >>> d = weakref.WeakKeyDictionary()

>       >>> d[k1] = 1   # d = {k1: 1}

Author: Maybe this could instead be 

```
# d = {id(k1): 1}
# d = {id(k1): 2}
```

Reviewer: That is not true I think. The key of `d` is `k1`, not `id(k1)`

Author: Ah yeah you're right, nevermind me. I was thinking of a way to emphasize that the ids are the same but this isn't quite right

   application without adding attributes to those objects.  This can be especially
   useful with objects that override attribute accesses.

   Note that when a key with equal value to an existing key (but not equal identity)
   is inserted into the dictionary, it replaces the value but does not replace the
   existing key. Due to this, when the reference to the original key is deleted, it
   also deletes the entry in the dictionary::

      >>> class T(str): pass
      ...
      >>> k1, k2 = T(), T()
      >>> d = weakref.WeakKeyDictionary()
      >>> d[k1] = 1   # d = {k1: 1}
      >>> d[k2] = 2   # d = {k1: 2}
      >>> del k1      # d = {}

   A workaround would be to remove the key prior to reassignment::

      >>> class T(str): pass
      ...
      >>> k1, k2 = T(), T()
      >>> d = weakref.WeakKeyDictionary()
      >>> d[k1] = 1   # d = {k1: 1}

>       >>> d.pop(k1)

Reviewer 2: ```suggestion
      >>> del d[k1]
```
This is more idiomatic if we don't care about the value associated with the removed key.

>    A workaround would be to remove the key prior to reassignment::
> 
>       >>> class T(str): pass
>       ...
>       >>> k1, k2 = T(), T()
>       >>> d = weakref.WeakKeyDictionary()
>       >>> d[k1] = 1   # d = {k1: 1}
>       >>> d.pop(k1)
>       >>> d[k2] = 2   # d = {k2: 2}
>       >>> del k1      # d = {k2: 2}
> 

Author: If less noisy is preferred:

```suggestion
   To workaround this, simply remove the key prior to reassignment.

```

Reviewer 2: I think this is confusing enough that the longer example is helpful.
